### ISSUE TITLE
File output to save potentiometer offset

### ISSUE THREAD

**Issue by ddsnowboard** [RoboRebel/RoboRebels2016#1]
Whenever we change anything about the potentiometer, all our pot code is going to break. We need to make something that will save a zero point without having to write code. What I propose, and what @dfuglsang and Mr. Holland agree with me on, is that we have a SmartDashboard button that we use to set the zero point in a text file on the RoboRio. 
Imagine the chain on the pot breaks, and we have to put it back on in the pits. We would boot it up in the pits, move the arm to the floor, and press the "Zero Potentiometer" button on the SmartDashboard. That will write out a text file on the RoboRio with the zero point in it. Then, we turn off the robot like usual. When we turn it on next, and every other time we turn it on, the robot reads the text file and sets the zero point of the potentiometer from it. This way, we can know the starting position of the potentiometer without a limit switch and without having to push code every time something changes. 

According to the Internet, there is already a built-in file format and class in Java for properties files (https://docs.oracle.com/javase/tutorial/essential/environment/properties.html). We can just put a call to this into the `Robot` constructor or something, setting the zero point of the pot each time. Then we have a button on the SmartDashboard that writes to this file. That should be it.

**Comment by dfuglsang** [comment]
Sounds like a good plan Will.  If it would help, I committed a
PropertiesUtils class to RoboRebels2016 GitHub repository that can assist
in reading and writing .properties files.  As you mentioned we could load
the properties in the Robot.robotInit() method and use a SmartDashboard
button to save/update that properties file. Maybe each subsystem can have a
getProperties():Properties or a updateProperties(Properties):void method
that Robot calls to gather the properties from each subsystem that we might
want to persist.

**Comment by ddsnowboard** [comment]
We made the feature detailed in the original post work in https://github.com/RoboRebel/RoboRebels2016/commit/8e873dcb8e4fc5ac0fc3522e1f11ad4e10505e46, but Mr. Holland suggested some changes such that you set both the high and low points. You would first put the arm on the ground and set the low point. By doing this, you also set the high point to a special value (-1, perhaps) that indicates to the software that the low point has just been set and the high point is ready to be set, but is meaningless in a physical context. Then you set the high point. If you try to set the high point first, it will tell you that you have to set the low point first and not do anything. 

This will make things even safer, because if you set the low point wrong with the original setup, you could still go too far and break the robot. With this setup, setting one point wrong will leave the other one still functional, and only setting one of the points will make the robot throw an error and not do anything, which would at least be safe.

This also prevents us from having to rewrite a bunch of the `BallHolderSubsystem` code to use a zero-point and a range, instead allowing us to use the old code that just had a top and a bottom end. We have to make sure we know what direction the pot goes when we move the arm up and down because that will affect how those variables are used, but we can't do that until we have an arm, so I'll worry about it then. 

Also, when we put this code on the other robot, there won't be a text file on it to begin with, so it will get mad. We have to include some code that creates a file and puts in a placeholder value if it can't find anything.

**Comment by dfuglsang** [comment]
I like the idea of having two Smart Dashboard buttons, one to set the max
angle and one to set the min angle.  You should think about the use of a
special value (e.g. -1) to dictate the behavior when resetting the min/max
values.  I would be worried about the "If you forget to set one with this
setup, the robot won't do anything" behavior,. If somehow only one of the
values gets reset between matches this behavior would result in our robot
not being able to shoot during a match.   After setting the min/max values
initially at school it may be better to only overwrite only the value
mapped to that particular reset button.  Something to think about.

**Comment by ddsnowboard** [comment]
@dfuglsang Do you think the special value would be a good or bad idea? And the robot could still shoot if it were set wrong, but you'd have to use the force button to move. If we just let each button overwrite its own number and have no effect on the other, the robot would be able to tear itself apart if it was set wrong.

**Comment by dfuglsang** [comment]
The special value may work but it would be useful to talk through all the
possible scenarios to make sure we don't cripple our robot in any way.
IMO, the most important value is the lower(?) one associated with the
maximum rotation of the ball holder forward since this would be used to set
the proper position for picking up balls - this is especially important
when the driver cannot see the robot in order to manually adjust the height
of the ball holder.  The upper(?) value, associated with rotating the ball
holder back to shoot at the upper goal, may not come into play unless we
are trying to shoot the ball straight up in the air for some reason.

It is important that we have safely features, either through our software
or hardware, to prevent our robot from tearing itself apart.  We still may
want to consider installing limit switches as a backup mechanism to stop
the ball holder from rotating too far in either direction.

**Comment by ddsnowboard** [comment]
You're right, we should make sure that the special value is handled in a logical way. There used to be a method called something like `isInRange()`, and I used to call that to check if the arm could move. What we could do is have two similar methods, `canRaise()` and `canLower()`, that look at the values that we have set, and do what they sound like if the values are real, and do something logical but controlled if they are not. Then we could use these methods instead of checking manually in commands et cetera. 

As I thought about it, preventing the robot from breaking itself was the main goal of this feature, more so than automatic aiming. The values that this feature will read are meant to be physical limits, past which the material structure of the robot will not allow us to go. Based on this, I'd say that the upper value is more important because it would prevent the electronics board from being broken. 

Ideally, we would also have limit switches to supplement or maybe replace this functionality, but I don't want to assume that we'll have them when it's not a sure thing. 

I think that the automatic aiming should be a separate feature from this because this is set up so that it can't be bypassed unless the driver really wants to and knows what he is doing. Automatic aiming doesn't need this. Automatic aiming would also have fewer safeties in the setting logic because no physical damage would result from doing it wrong.
